liblttng-ust-java-agent/java/lttng-ust-agent-common/org/lttng/ust/agent/AbstractLttngAgent.java

```diff
@@ -52,29 +50,22 @@
 	/**
 	 * The trace events currently enabled in the sessions.
 	 *
-	 * The key represents the event name, the value is the ref count (how many
-	 * different sessions currently have this event enabled). Once the ref count
-	 * falls to 0, this means we can avoid sending log events through JNI
-	 * because nobody wants them.
+	 * The key is the {@link EventNamePattern} that comes from the event name.
+	 * The value is the ref count (how many different sessions currently have
+	 * this event enabled). Once the ref count falls to 0, this means we can
+	 * avoid sending log events through JNI because nobody wants them.
 	 *
 	 * It uses a concurrent hash map, so that the {@link #isEventEnabled} and
 	 * read methods do not need to take a synchronization lock.
 	 */
-	private final Map<String, Integer> enabledEvents = new ConcurrentHashMap<String, Integer>();
+	private final Map<EventNamePattern, Integer> enabledPatterns = new ConcurrentHashMap<EventNamePattern, Integer>();
 
 	/**
-	 * The trace events prefixes currently enabled in the sessions, which means
-	 * the event names finishing in *, like "abcd*". We track them separately
-	 * from the standard event names, so that we can use {@link String#equals}
-	 * and {@link String#startsWith} appropriately.
-	 *
-	 * We track the lone wildcard "*" separately, in {@link #enabledWildcards}.
+	 * Cache of already-checked event names. As long as enabled/disabled events
+	 * don't change in the session, we can avoid re-checking events that were
+	 * previously checked against all known enabled patterns.
 	 */
-	private final NavigableMap<String, Integer> enabledEventPrefixes =
-			new ConcurrentSkipListMap<String, Integer>();
-
-	/** Number of sessions currently enabling the wildcard "*" event */
-	private final AtomicInteger enabledWildcards = new AtomicInteger(0);
+	private final Map<String, Boolean> enabledEventNamesCache = new HashMap<String, Boolean>();
```

Comment:
we're doing a non-synchronized get in a non-concurrent hashmap in the "is enabed" check. Is it buggy ?